lib/ansible/modules/cloud/vmware/vmware_guest_powerstate.py

```diff
@@ -0,0 +1,160 @@
+    username: administrator@vsphere.local
+    password: vmware
+    validate_certs: no
+    folder: /testvms
+    name: testvm_2
+    state: poweredoff
+  delegate_to: localhost
+  register: deploy
+'''
+
+RETURN = r''' # '''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.vmware import (HAS_PYVMOMI, connect_to_api, set_vm_power_state, find_vm_by_id,
+                                         vmware_argument_spec)
+
+
+class PyVmomiHelper():
+    def __init__(self, module):
+        if not HAS_PYVMOMI:
+            module.fail_json(msg='PyVmomi Python module required. Install using "pip install PyVmomi"')
+
+        self.module = module
+        self.params = module.params
+        self.si = None
+        self.content = connect_to_api(self.module)
+
+    def get_vm(self):
+        vm = None
+        match_first = True if self.params['name_match'] == 'first' else False
+
+        if self.params['uuid']:
+            vm = find_vm_by_id(self.content, vm_id=self.params['uuid'], vm_id_type="uuid")
+        elif self.params['folder'] and self.params['name']:
+            vm = find_vm_by_id(self.content, vm_id=self.params['name'], vm_id_type="inventory_path",
+                               folder=self.params['folder'], match_first=match_first)
+        return vm
+
+
+def main():
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(
+        state=dict(type='str', default='present',
+                   choices=['poweredoff', 'poweredon', 'rebootguest', 'restarted', 'shutdownguest', 'suspended']),
+        name=dict(type='str'),
+        name_match=dict(type='str', choices=['first', 'last'], default='first'),
+        uuid=dict(type='str'),
+        folder=dict(type='str', default='/vm'),
+        force=dict(type='bool', default=False),
+    )
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=True,
+                           required_one_of=[
+                               ['name', 'uuid'],
+                           ],
+                           )
+
+    result = {'failed': False, 'changed': False}
+
+    pyv = PyVmomiHelper(module)
+
+    # Check if the VM exists before continuing
+    vm = pyv.get_vm()
+
+    if vm:
+        # VM already exists, so set power state
+        tmp_result = set_vm_power_state(pyv.content, vm, module.params['state'], module.params['force'])
+        if tmp_result['changed']:
+            result["changed"] = True
+            result["instance"] = tmp_result['instance']
+        if not tmp_result["failed"]:
+            result["failed"] = False
+    else:
+        # If UUID is set, getvm select UUID, show error message accordingly.
+        msg = "Unable to set power state for non-existing virtual machine : "
+
+        if module.params['uuid']:
+            msg += module.params['uuid']
+        else:
+            msg += module.params['name']
+
+        module.fail_json(msg=msg)
+
+    if result['failed']:
```

Comment:
Replace this with:
```python
if result.get('failed'):
    module.fail_json(**result)

module.exit_json(**result)
```

Reads a lot easier.

Reply:
Ok.

Reply:
OK.